lib/utils/fncache_test.go

```diff
@@ -178,14 +181,21 @@ func TestFnCacheCancellation(t *testing.T) {
 	// unblock the loading operation which is still in progress
 	close(blocker)
 
-	ctx, cancel = context.WithTimeout(context.Background(), timeout)
+	// use long timeout since we need to wait for some concurrent operations to complete
+	// behind the scenes, and that can take a while in a resource-constrained environment.
+	ctx, cancel = context.WithTimeout(context.Background(), longTimeout)
 	defer cancel()
 
+	loadFnWasRun := atomic.NewBool(false)
 	v, err = cache.Get(ctx, "key", func(context.Context) (interface{}, error) {
-		t.Fatal("this should never run!")
+		loadFnWasRun.Store(true)
 		return nil, nil
 	})
 
+	if loadFnWasRun.Load() {
```

Comment:
Checking the status here doesn't create the race condition? We check the condition, and the function above runs after?

Reply:
The previous call to `Get` returns for one of 3 cases:

- We loaded the already cached value (success).
- We hit the context timeout (failure, checked below).
- Load function was run to completion before the timeout was hit (failure, checked here).

So it would be racy if we didn't check for the timeout, but since we do its safe.